[Proposal] Make mouseWheel zoom in % instead of pixels value

RomanJos opened the pull request:
## Description
I noticed that InteractiveWidget scrolled by a certain amount of px instead of a percentage, thus using a big child required more scrolling so I used a fixed value of 0.1 scale units (don't know how much % it is) per scroll instead of the previous px amount of scroll divided by height.

I don't know if the previous code was related to this setting on Windows, if yes then my code ignore that :
![image](https://user-images.githubusercontent.com/42180076/100339935-c934f880-2fda-11eb-9fa5-b5f0a4d273e8.png)


You can see the before and after :
![InteractiveViewer Before](https://user-images.githubusercontent.com/42180076/100340959-38f7b300-2fdc-11eb-9d5a-4befb0a64bc2.gif)
![InteractiveViewer After](https://user-images.githubusercontent.com/42180076/100340961-39904980-2fdc-11eb-8b44-20cb7659c374.gif)



## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

RomanJos committed:
Make mouseWheel zoom in % instead of pixels value

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -932,10 +932,12 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
       final Size childSize = childRenderBox.size;
-      final double scaleChange = 1.0 - event.scrollDelta.dy / childSize.height;
-      if (scaleChange == 0.0) {
+      //Ignore left and right scroll
```
We put a space after the `//` and end comments with a period.  See the [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) for more.

Same for the other comment below.

justinmc reviewed (commented):
Thanks for submitting this PR.  You've got a point that the scrolling calculation might not be ideal for certain situations, but I don't think that the way you have it now is exactly right either.

In this example I tried, this PR seemed to make the zooming much too fast:

| Current | This PR |
| ---- | ---- |
| ![scroll_original](https://user-images.githubusercontent.com/389558/101090761-fef25800-356b-11eb-852f-9c01badbbd9e.gif) | ![scroll_pr](https://user-images.githubusercontent.com/389558/101090773-031e7580-356c-11eb-8687-568ae5868615.gif) |

Can you come up with any way to make it work nicely for all cases?

If you are able to figure that out, this PR will also need a test, by the way.  You can add it to interactive_viewer_test.dart.

*Update*: Here's the code from the gif: https://github.com/justinmc/flutter-interactive-viewer-demos/blob/master/lib/image/image_page.dart

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -932,10 +932,12 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
       }
       final RenderBox childRenderBox = _childKey.currentContext!.findRenderObject()! as RenderBox;
       final Size childSize = childRenderBox.size;
```
This variable childSize would no longer be used with these changes and could be removed.

RomanJos commented:
Hello, would it be possible to have the code you used in your gif ? At first I thought my changes worked only when constrained was set to false but it work in both cases

justinmc commented:
Yes, sorry for the late reply.  Here is the code: https://github.com/justinmc/flutter-interactive-viewer-demos/blob/master/lib/image/image_page.dart

That's my repo of InteractiveViewer demos, let me know if you can't get it running.

RomanJos committed:
Update interactive_viewer.dart

RomanJos commented:
Oh I know exactly whats going on, you use a trackpad while I use a mouseWheel, trackpad send different amount of pixels than the mouseWheel which is hardcoded to 20 here :
https://github.com/flutter/engine/blob/d2ccdefe1877ca46e389bf440ea3c13c7bcfbf5c/shell/platform/windows/win32_flutter_window.cc#L18-L20
> Now I know the windows mouse settings is completly ignored by the Flutter engine

Mannualy set 1.1 scale where it should be 0.01 for example is why it zoom too much on your gif, my new solution is to divide the scroll amount by 200 so a scroll of 20 give 0.1 but a scroll of 2 give us 0.01.

RomanJos committed:
Use exp()

I noticed in the test that scrolling two time a value and scrolling back one time 2*value wasn't resetting the zoom to 1.0 so using exp fix it

RomanJos committed:
Update interactive_viewer_test.dart

RomanJos commented:
Hello, I think I fixed everything! I tried my best for the test even thought I didn't really know what to do but scrolling with the mouse now work well on desktop as well as trackpad. Using exp allow to avoid scaling to or past 0.0 in every way so thats great

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -930,12 +930,16 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
+
+      // The mouseWheel scroll amount is hardcoded to 20 while a trackpad scroll can be any amount
+      // by dividing the scroll amount by 200 it scale 0.1 units per mouseWheels scroll and feel natural for trackpads.
+      // See https://github.com/flutter/flutter/pull/71266
```
Nitpick: Some slight rewording. I also think it's not necessary to link the PR, since it can be found through git.

In the Flutter engine, the mousewheel scrollDelta is hardcoded to 20 per scroll, while a trackpad scroll can be any amount.  The calculation for scaleChange here was arbitrarily chosen to feel natural for both trackpads and mousewheels on all platforms.

RomanJos replied:
Nice rewording !

justinmc commented on packages/flutter/lib/src/widgets/interactive_viewer.dart:
```diff
@@ -930,12 +930,16 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
+      // The mouseWheel scroll amount is hardcoded to 20 while a trackpad scroll can be any amount
+      // by dividing the scroll amount by 200 it scale 0.1 units per mouseWheels scroll and feel natural for trackpads.
+      // See https://github.com/flutter/flutter/pull/71266
+      final double scaleChange = math.exp(-event.scrollDelta.dy / 200);
```
Good idea using e.  I tried this again on my Macbook's trackpad and also using a Magic Mouse, and both felt natural to me.

justinmc commented on packages/flutter/test/widgets/interactive_viewer_test.dart:
```diff
@@ -672,6 +672,37 @@ void main() {
+      await scrollAt(center, tester, const Offset(0.0, -200.0));
+      await tester.pumpAndSettle();
+      // math.exp round the number too short compared to the one in transformationController.
+      expect(double.parse((transformationController.value.getMaxScaleOnAxis()).toStringAsFixed(15)), math.exp(400 / 200));
```
I think you can use [closeTo](https://master-api.flutter.dev/flutter/package-matcher_matcher/closeTo.html) here instead of converting to String.

RomanJos replied:
I put the max digit after 0 just before it break

justinmc reviewed (commented):
Thanks for following up on this.  I think this looks like a good improvement to me.  I tried this with a trackpad and a Magic Mouse for both big and small children, and I set the min/max scales to allow zooming very very far, and the zooming felt natural to me at all times 👍 

I left some comments about improving the docs and fixing the analyzer errors, then I think this is good to go.

RomanJos committed:
Update interactive_viewer_test.dart

RomanJos committed:
Update interactive_viewer.dart

justinmc commented:
There are some analyzer errors:

```
  error • Undefined name 'math' • packages/flutter/test/widgets/interactive_viewer_test.dart:696:66 • undefined_identifier
  error • The argument type 'dynamic' can't be assigned to the parameter type 'num' • packages/flutter/test/widgets/interactive_viewer_test.dart:700:74 • argument_type_not_assignable
  error • Undefined name 'math' • packages/flutter/test/widgets/interactive_viewer_test.dart:700:74 • undefined_identifier
```

RomanJos commented:
oops I forgot to import it in the commit !

RomanJos committed:
Update interactive_viewer_test.dart

justinmc commented:
The test failures are something to do with SVG, which I have seen before on one of my PRs and I think is unrelated to this.  Can you try merging master into this and see if that makes it pass?

RomanJos committed:
Merge branch 'master' of https://github.com/flutter/flutter into patch-1

justinmc reviewed (approved):
LGTM 👍 

The tests are passing so I'll go ahead and merge this.  Thanks for your work on this!

RomanJos commented:
Great ! I'm glad I made Flutter better, thank you Justin for your effort and support !
